### DISCUSSION TITLE
Modified script for fully automated FNIS conversion

### DISCUSSION THREAD

**Discussion by lobotomy-x** [Monitor221hz/Pandora-Behaviour-Engine-Plus#39] [Patches]
Couldn't figure out how the posted script was intended to be used for batch processing so I modified it to be usable as an MO2 executable. Place the files in your pandora folder (should be a mod in mo2 so that all the .exe files are in your base level data folder in the Vfs) and add the bat file as an MO2 executable with a proper output folder. It will read any behavior file you have with FNIS_ in the name and attempt conversion with hkxconv for 64 bit files and hkxcmd for 32 bit which was crucial for me as I have previously converted all of my behavior files to 64 bit without keeping the originals. I also added a command to hkxconv to suppress warnings and force conversion for the outdated file types. It will generate files but doesn't seem to work in game from my limited testing.

Important to note that I chose to display error messages individually for the 32 bit and 64 bit attempts so it's normal to see lots of errors. There's only a concern when both versions fail. I'll try to adjust that to be a bit more intuitive. You could add powershell as an mo2 executable and from there launch the patcher so you can see the window as it runs, otherwise there is a log file but it currently only prints the errors. It will not clear previous runs and rather appends to the same log file. Delete it if you want a fresh log, otherwise scroll down. On repeated runs you should see a message that files already exist at the bottom of your log. 

Originally I had each behavior processed as a new mod which resulted in an obscene amount of mods so now there is some parsing to group stuff together. Its pretty bad as I have limited python experience and had to ask Bard to help me with that part. Ideally it would be great to make this into an mo2 plugin that can query the meta files to get author, name, and link or at least better process the names. We're also reaching the point the main pandora patcher probably needs a shift-click function to select multiple patches.

Be aware this will try to convert XPMSE so delete the generated mod if you're using the provided pandora XPMSE mod. There's probably tons of bugs and you will need to examine the output in all likelihood but this should be an improvement for most people. I was honestly shocked but very pleased when I got in game and almost everything worked just as well as my previous nemesis/fnis setup. We're very close to the all in one dream now!

I don't need any credit if you modify this script. Please do so if you are capable as it could use some work and getting it to this point already gave me brain damage. I will say that it's been much more difficult to get a complete patch from Pandora lately. I'm unsure if that's due to changes in the main patcher handling conflicts differently with my set of mods or if these fnis mods conflict with my combat mods, but do be aware this is still a prerelease mod. Using it as a nemesis replacer is probably not wise if you can't also successfully patch nemesis with the same mods. 

[Patcher.zip](https://github.com/Monitor144hz/Pandora-Behaviour-Engine-Plus/files/13758824/Patcher.zip)

**Comment by Kyationa** [comment]
so i just unzip and put those into to pandora folder do i make a folder in pandora folder or no?

**Reply by Kyationa** [reply]
i cant get this to work could i get a screen shot of what i need to do if possible?

**Reply by Kyationa** [reply]
seems to fail at every single one is this what it suppose to do?

[log_file.txt](https://github.com/Monitor144hz/Pandora-Behaviour-Engine-Plus/files/13765126/log_file.txt)

**Reply by Kazyachka** [reply]
@elbadcode Hello, this is strange, but I get an instant crash, I can’t even imagine what this might be connected with

**Reply by lobotomy-x** [reply]
Are you sure it crashed? Check the output folder you should've specified one in mo2. It may have the finished patch. The log file doesn't show anything unusual as I didn't bother to disable the error messages when either 32 bit or 64 bit conversion fails. It would only be an issue if it fails both. You can see the window while it works if you run it from powershell or terminal but as long as you have the necessary files in the same folder it should run in a minimized window after you hit start. You should also run mo2 as admin so the script will run as admin. And don't attempt this on a Skyrim installed in the steam library, you should use a different install location for Skyrim anyway.


And again just unzip these into your Pandora 0.8.4 folder along with hkxconv and hkxcmd which you will need to go obtain yourself. That folder should be installed as a mod in mo2 so that essentially you are running this straight from the data folder in the mo2 virtual filesystem

If you do all of that and show that your output and mo2 overwrite folder are still empty I'll help you in more detail and see if there's a bug in the script. Otherwise you should continue to be patient as monitor works on the main patcher and I will likely update this patcher to be more user friendly simply for my own purposes. This kind of prerelease stuff is mainly intended for people who can DIY a bit. And although my patch worked faster, it still has a bit less mod coverage than the old tried and true method of running fnis and then disabling it and running nemesis.

**Reply by Kyationa** [reply]
so i got this far using this is this how it is suppose to do it?
[log_file.txt](https://github.com/Monitor144hz/Pandora-Behaviour-Engine-Plus/files/13767420/log_file.txt)

**Reply by lobotomy-x** [reply]
at the bottom it says found existing files. that means it worked and you should see the mods in pandora

**Reply by Kyationa** [reply]
the only ones that don't seem to load are the ones with Babo in the name any idea of a fix that i could try?

**Reply by lobotomy-x** [reply]
I'd need to test those Babo == baka, right? It may be due to issues with my parsing script not producing the files at all or it may be that they have the old formatted FNIS files that hkxconv doesn't work with. I'm still learning about behavior modding and which tools work for modern Skyrim since info is all over the place but it would be nice to find a resolution on those. For those specific mods if I don't find a resolution it may be worth asking the creator directly

**Reply by Kyationa** [reply]
yeah there the baka mods i tried a few thing it can make a ini  and hkx file but it wont do anything from there.

**Reply by Kyationa** [reply]
seems like for the most part things work just fnis animations are stuck in each other and i can't go back to nemesis cause it gives me a error missing meshes in the skyrim folder.

**Reply by lobotomy-x** [reply]
Simply reinstall fnis/nemesis and it should fix the missing behaviors error. Unsure what you mean by fnis animations are stuck in each other. I found everything except a few animations worked as normal and the non working ones simply didn't play

**Reply by Kyationa** [reply]
the characters stay inside each other and i tried the reinstall and it still gave me errors tried a fresh profile should i manually install nemesis and see if that works if i cant fix the animation issues?

**Reply by Kyationa** [reply]
ERROR(1210): Vital behavior file cannot be found. This will cause T-pose in game. To fix this, either reinstall the mod or manually restore this behavior file
File: d:\steam\steamapps\common\skyrim special edition\data\meshes\actors\bear\behaviors\quadrupedbehavior.hkx
 this is the error with nemesis im keep testing pandora.

**Reply by Kyationa** [reply]
the naming of the folders seems to name them incorrectly with names infront of the standard 0_master and the defaultfemale and the defaultmale i also feel the creature file names might be incorrect but i don't understand how those would be named so it reads it correctly

**Reply by lobotomy-x** [reply]
The names of the behavior folders are correct and consistent with Pandora format which is different than you'd be used to with nemesis. I didn't really touch that part anyway. The naming of the mods is incorrect often yes but you can just run this once and manually group up generated mods into the same Pandora mod for easier patching

**Reply by Kyationa** [reply]
ok do you have the way the creature mods naming for files should look like wanna make sure i have it correct in my list i fixed the character mod names i just need the how they are name in file format the wiki for creature naming makes less sense with the creature

**Reply by Kyationa** [reply]
also do we just keep the Pandora folder with everything in it or do we keep all things generated by the converter being the hkx and xml that are outside the Pandora folder
